src/pool.ts

```diff
@@ -0,0 +1,134 @@
+ * concurrent operations.
+ */
+export class ClientPool<T> {
+  private activeClients: Map<T, number> = new Map();
+
+  /**
+   * @param {number} concurrentOperationLimit - The number of operations that
+   * each client can handle.
+   * @param {() => T} clientFactory - A factory function called as needed when
+   * new clients are required.
+   */
+  constructor(
+      private readonly concurrentOperationLimit: number,
+      private readonly clientFactory: () => T) {}
+
+  /**
+   * Returns an already existing client if it has less than the maximum number
+   * of concurrent operations or initializes and returns a new client.
+   */
+  private acquire(): T {
+    let selectedClient: T|null = null;
+    let currentRequestCount = 0;
+
+    this.activeClients.forEach((requestCount, client) => {
+      if (!selectedClient && requestCount < this.concurrentOperationLimit) {
+        selectedClient = client;
+        currentRequestCount = requestCount;
+      }
+    });
+
+    if (!selectedClient) {
+      selectedClient = this.clientFactory();
+      assert(
+          !this.activeClients.has(selectedClient),
+          'The provided client factory returned an existing instance');
+    }
+
+    this.activeClients.set(selectedClient, currentRequestCount + 1);
+
+    return selectedClient!;
+  }
+
+  /**
+   * Reduces the number of operations for the provided client, potentially
+   * removing it from the pool of active clients.
+   */
+  private release(client: T): void {
+    let currentRequestCount = this.activeClients.get(client);
+    assert(currentRequestCount! > 0, 'Active client not found');
+
+    currentRequestCount = currentRequestCount! - 1;
+    this.activeClients.set(client, currentRequestCount);
+
+    if (currentRequestCount === 0) {
+      this.garbageCollect();
+    }
+  }
+
+  /**
+   * The number of currently registered clients.
+   *
+   * @return {number} Number of currently registered clients.
+   */
+  // Visible for testing.
+  get size(): number {
+    return this.activeClients.size;
+  }
+
+  /**
+   * Runs the provided operation in this pool. This function may create an
+   * additional client if all existing clients already operate at the concurrent
+   * operation limit.
+   *
+   * @param {(client: T) => Promise<V>} op - A callback function that returns a
+   * Promise. The client T will be returned to the pool when callback finishes.
+   * @return {Promise<V>} A Promise that resolves with the result of `op`.
+   */
+  run<V>(op: (client: T) => Promise<V>): Promise<V> {
+    const client = this.acquire();
+
+    return op(client)
+        .catch(err => {
+          this.release(client);
+          return Promise.reject(err);
+        })
+        .then(res => {
+          this.release(client);
+          return res;
+        });
+  }
+
+  /**
+   * Deletes clients that are no longer executing operations. Keeps up to one
+   * idle client to reduce future initialization costs.
```

Comment:
FWIW, I suspect we could get away without any intelligent shrinking of the pool...  clients probably aren't very expensive (especially since there are idle timeouts on the connections, right?)...  so if we wanted we could have a more naive pool implementation.  But since you've gone to the trouble and it's not too complicated, there's probably no reason not to proceed with your fancy resizing pool stuff. :-)

Reply:
This pool will likely be used by millions of other consumer in NPM land. I am sure that one of them will need this garbage collection. :)